src/Data/Array/Polarized/Push.hs

```diff
@@ -67,6 +74,21 @@ make x n
   | n < 0 = error "Making negative length push array"
   | otherwise = Array (\makeA -> mconcat $ Prelude.replicate n (makeA x))
 
+singleton :: a -> Array a
+singleton x = Array (\writeA -> writeA x)
+
+snoc :: a -> Array a %1-> Array a
+snoc x (Array k) = Array (\writeA -> (k writeA) <> (writeA x))
+
+cons :: a -> Array a %1-> Array a
+cons x (Array k) = Array (\writeA -> (writeA x) <> (k writeA))
+
+foldWith :: Monoid b => (a -> b) -> Array a %1-> b
+foldWith f (Array k) = k f
+
+unzip :: Array (a,b) %1-> (Array a, Array b)
+unzip = undefined
```

Comment:
```suggestion
unzip (Array k) = k (\(a, b) -> (singleton a, singleton b))
```

This kind of make sense to me, but I'm unsure (both semantically and performance-wise).

Reply:
That's *very* clever and I feel dumb for not coming up with it after messing with it for a while, lol :smile_cat: 

It does make sense: `k` is something that takes something that writes an `a` to a monoid that represents a composable way to write an `a` to a cell of an array. Here, `k` takes something that changes `(a,b)` to a monoidal-computation that writes an `(a,b)` to a cell of an array. 

This implementation in words says "a way to write each pair of (a,b) into a monoidal writing-computation is to have a pair of monoidal writing-computations".  If you give this to `k` it will make a bunch of pairs of singletons and concatenate them.